### ISSUE TITLE
[UX] Figure out a "better" way to enable/disable modules.

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#541]
This is a follow-up/spin-off from #501 to prevent derailing the actual goal of that issue.

At some point, it was pointed that the "Enabled" column of the modules page could be less wide. That suggestion gave birth to a lot of ideas with regards to how enabling/disabling modules could work and discussion seemed to be getting OT. Lets take it here and keep that issue clear.

---

Winning Ideas below:
* automatically select dependancies 
* Sticky submit button (at bottom of screen) 
* Change enabled/disabled to on/off [with more other stuff](https://github.com/backdrop/backdrop-issues/issues/541#issuecomment-491617869)

---

See also
* [Installing modules through config](https://github.com/backdrop/backdrop-issues/issues/435)
* [Streamline the process of disabling and uninstalling](https://github.com/backdrop/backdrop-issues/issues/2532)

**Comment by klonos** [comment]
@jenlampton:

> I love these recommendations. :) To make the enabled column narrower, how would you feel about having modules be "off" and "on" instead of "enabled" and "disabled" ?

**Comment by klonos** [comment]
I personally do not mind on/off instead of enabled/disable, but what actually makes the "Enabled" column as wide as it is is its title/heading. Changing the "Enabled" heading to something like "on/off" saves us a couple of characters, but would not make that much of a difference. If we choose to go that way, I prefer "Status" as a heading. It makes more sense to me.

**Comment by klonos** [comment]
...on second thought, would it make any sense to remove the "Enabled" column altogether and replace the "on/off" checkboxes with a "enable/disable" AJAX operation?

That would save us the whole width since for disabled modules the checkbox would be moved as a "enable" operation button in the already existing "Operations" column. For enabled modules, the "disable" operation would be "merged" with the drop-button as with any other module that has multiple operations available ("configure", "permissions").

**Comment by klonos** [comment]
@quicksketch:

>> ...on second thought, would it make any sense to remove the "Enabled" column altogether and replace the "on/off" checkboxes with a "enable/disable" AJAX operation?

>I think this may have a negative user experience impact. Every module enable/disable requires a full cache flush. On many sites I work on, enabling a module may take 10-20 seconds. If you need to turn on 6 modules, that's going to be a frustrating experience waiting for each AJAX request to finish before you can enable the next module.

>> Changing the "Enabled" heading to something like "on/off" saves us a couple of characters, but would not make that much of a difference.

>Good point here. Unless we renamed the column header to just "On", it wouldn't save hardly any space, and "On" just isn't an adequate descriptor.

**Comment by klonos** [comment]
Fair enough about the long time and thus bad UX to enable/disable using AJAX. I was just brainstorming.

How about if the "enable"/"disable" operations simply "stacked" modules for the respective action + it invoked a "dirty" flag? Then we could use the same mechanism that prompts the user to save before changes take effect like we do in views. As an added visual indication/effect we could have disabled modules name/version/description be grayed out and enabled ones be normal.

So, hitting the "enable" button on disabled/grayed out modules would:

    [UI] "light up" the module font
    [UI] turn the "enable" button text to "disable"
    [back end] queue the module for enabling
    [back end] set a "dirty" flag for the modules page
    [UI] fire a "All changes are stored temporarily. Click Save to make your changes permanent." message

...and similarly hitting the "disable" action for enabled modules would:

    [UI] "dim" the module font
    [UI] turn the "disable" button text to "enable"
    [back end] queue the module for disabling
    [back end] set a "dirty" flag for the modules page
    [UI] fire a "All changes are stored temporarily. Click Save to make your changes permanent." message

This way, the user can keep on enabling and/or disabling modules and things happen only when they hit the "save" button.

**Comment by klonos** [comment]
..anyways, my previous comment might be the means to and end, but is totally off-topic here. If you think that what I propose might improve UX, we could move that to its own issue.

For now, the whole idea is to figure out if we could remove the "Enabled" column in order to save horizontal space before trying to align columns so that they have equal widths throughout the page.

**Comment by klonos** [comment]
@BWPanda:

> Here's another idea: make the Modules page the same as other admin pages (e.g. admin/content) - the checkboxes column doesn't have a label, they're just for selecting items to apply an action to. In this case, the enabled/disabled status of a module could be based on the color of the table row (green for enabled, red or grey for disabled, etc.) and you simply check the boxes you want to enable then apply the 'Enable' action (bulk operation). You could go one step further and add an 'Uninstall' action so that you could uninstall disabled modules from the one page (rather than going to a separate tab - disabled modules would have to be colored differently to uninstalled modules). You could even add a 'Disable & Uninstall' action to do both in one go. Here's what it might look like:

> ![modules](https://cloud.githubusercontent.com/assets/2385329/5605164/44a3035a-9440-11e4-8847-42904201a0a9.jpg)

> Otherwise, I don't really see a problem with the width of the 'Enabled' column (but then I only ever view the admin pages on my widescreen desktop). The solution is to change the auto-column-width functionality to make that column (and the 'Operations' column) a fixed-width instead of a variable-width.

**Comment by klonos** [comment]
++ for no label!!

...not sure about bulk operations as a drop-down + "execute" button though.

**Comment by docwilmot** [comment]
Two hacks would please me. One if dependencies were automatically selected (maybe with a fading JavaScript pop up to warn you), and if the save button could float above the page, probably over the floating tableheader. Failing that an anchor link that takes you to the save button. And maybe a save button at the top of the page. On mobile especially, scrolling to the save button is tedious.
I'm not sure about the ajaxy button enabling either unless it actually did enable, which would as @quicksketch said slow things down. This would also be the same UX process as currently done - select module, find save button, click it - except with an Ajax button instead of a checkbox, and more complicated.
I don't think I like the VBO style for reasons @klonos said, plus sometimes if I'm just enabling one module it's nice to just click the checkbox and hit enter. That would be lost.

**Comment by jenlampton** [comment]
+++ for no label!!
And, that's an easy win. (quick PR please?)

I also like the idea of dependencies being automatically selected via JS.

We have problems with the location of the submit button in views, too.  It's the only UI in core that puts it at the top right, and if your live preview is too wide, that submit button scrolls right off the page (scrolls to the right - where no one can find it).  Here's an interesting idea, you know how we have a fancy new sticky-footer in the Bartik theme? Why don't we create a similar fancy new sticky-footer for forms where we can stash those submit buttons? This could be a handy UI pattern for any form that gets too long. The buttons would always be at the bottom left of the screen, but let the page scroll behind.

**Comment by docwilmot** [comment]
> Why don't we create a similar fancy new sticky-footer for forms where we can stash those submit buttons? The buttons would always be at the bottom left of the viewport, but we can let the page scroll behind.

Genius! +++

**Comment by klonos** [comment]
> I like the idea of dependencies being automatically selected via JS...

#565

> ...you know how we have a fancy new sticky-footer in the Bartik theme? ...

If you're talking about #473, then that's not a sticky footer per se. That commit made it so that **if the page content is shorter than the page height**, then the footer sticks at the bottom of the page (so not to leave blank/white space after the footer). What we need is an "always in-view" footer (or block/region). If we had it so that the admin menu was sticky on top, we could place form submit buttons (and perhaps "dirty" flag messages) there. But then again, not all users have access to the admin menu, so perhaps something like [Sticky Edit Actions](https://www.drupal.org/project/sticky_edit_actions): #566

**Comment by jenlampton** [comment]
Okay, so maybe it's not the same thing, but I like the idea of a sticky-button area, so save/submit/whatever is always at the bottom left of your screen. Consistency++

**Comment by BWPanda** [comment]
Since the title of this issue is 'Figure out a "better" way to enable/disable modules', let me suggest this:

Do what Drupal 8 has done (and probably most other CMS's (haven't checked)) and remove the distinction between enabled/disabled and installed/uninstalled modules - just have modules be either on (enabled/installed) or off (disabled/uninstalled).

I see it working like this (and I'll use the terms 'on' and 'off' for lack of better terminology): tick boxes next to modules that are off to turn them on. Tick boxes next to modules that are on to turn them off. On the confirmation page have two buttons - the default button turns the module off but keeps settings/data intact in case you want to turn it back on later without losing anything. The second button turns the module off and deletes all settings/data.
Additionally, there'd be a new page under Reports that lists all modules that are turned off but that still have settings/data in the system. From here you can delete that data, for when you didn't do that when turning the module off but now want to clean up your site from old config stuff.

Does that make sense? It'd be a big change, but I think it'd simplify the process for users wondering what the difference between disabled and uninstalled is.

**Comment by jenlampton** [comment]
@BWPanda I think we have a separate issue for that somewhere...  I can't find it, but will link a few related issues below.

> Do what Drupal 8 has done 

This was the one change in Drupal 8 that I fought the hardest against. It was the straw that broke my camels back, and made me believe we needed to fork. This one move basically said "We care nothing about our users or their data, and only about our developers." So this idea gives me the heebie-geebeies. However... :)

> On the confirmation page have two buttons - the default button turns the module off but keeps settings/data intact in case you want to turn it back on later without losing anything. The second button turns the module off and deletes all settings/data.

Oh how I wish this thought had occurred to ANYONE in Drupal 8! I'm so glad you're with us @BWPanda :)

> Additionally, there'd be a new page under Reports that lists all modules that are turned off but that still have settings/data in the system. 

Or an entry in the system table? (Even if the module didn't provide any data?)

> From here you can delete that data, for when you didn't do that when turning the module off but now want to clean up your site from old config stuff.

I wonder if there's a way we can expand this to include modules that have left their data behind, but also been removed from the filesystem? 

For example, if there's an entry in the system table, but no code in the file system, we could disable the "Delete all the data" button (or whatever it says) and add a "?" that explains that version X of this module will need to be added back to the site before data can be safely cleaned up.

> Does that make sense? It'd be a big change, but I think it'd simplify the process for users wondering what the difference between disabled and uninstalled is.

yes, and I love it :) I love the simplicity of on/off and moving the complex "installed/uninstalled" stuff to a dedicated report. 


See 
* [Installing modules through config](https://github.com/backdrop/backdrop-issues/issues/435)
* [Streamline the process of disabling and uninstalling](https://github.com/backdrop/backdrop-issues/issues/2532)

**Comment by BWPanda** [comment]
Aww thanks! I had that idea for offering to delete data when uninstalling something from, I believe, Linux

**Comment by BWPanda** [comment]
> I wonder if there's a way we can expand this to include modules that have left their data behind, but also been removed from the filesystem?

Yes, that's what I was envisioning too. Probably a whole separate issue, but I think there should be a way for core to clean up after modules so that if they're removed, they can still be cleaned up without having to reinstall them (having this problem on D8 lately when moving config between dev and live and modules have been removed but need to put them back just so config can be updated - argh!).

**Comment by klonos** [comment]
>>On the confirmation page have two buttons - the default button turns the module off but keeps settings/data intact in case you want to turn it back on later without losing anything. The second button turns the module off and deletes all settings/data.
>
> Oh how I wish this thought had occurred to everyone for Drupal 8! I'm so glad you're with us @BWPanda :)

Yup ... @BWPanda 👍 ❤️ 

>>Does that make sense? It'd be a big change, but I think it'd simplify the process for users wondering what the difference between disabled and uninstalled is.
>
>yes, and I love it :) I love the simplicity of on/off and moving the complex "installed/uninstalled" stuff to a dedicated report.

Don't you just love it when brainstorming, but doing so with the end user in mind, brings such wonderfull ideas to light? I love these ideas too!

> For example, if there's an entry in the system table, but no code in the file system, we could disable the "Delete all the data" button (or whatever it says) and add a "?" that explains that version X of this module will need to be added back to the site before data can be safely cleaned up.

Cross-linking the related #1624

If we implemented confirmation pages as confirmation dialogs instead (#3769), the UX would be improved so much more (no page reloads etc.)

**Comment by klonos** [comment]
...one thing to consider if we convert the checkboxes we currently have in place right now, into on/off buttons is that we'd lose the ability to bulk select modules for enabling/disabling. We could have a "enable/disable multiple modules" toggle link at the top-right (a-la "show row weights" for sortable tables, and "Manual entry" for the new options form element).

**Comment by docwilmot** [comment]
Its not clear to me what this would achieve though. In this scenario, the modules page would list all modules in the file system, and those that were "active" would have a checkbox next to them checked, and there would be a separate page which lists "inactive" modules.

Well this is exactly what we have now.

I believe the D7 killing of "disabling" was more to do with the fact that the idea that modules could be disabled and then safely re-enabled at a later stage was fallacious, in that the rest of the site data may  have moved on and may conflict with this "re-entry" data. (My paraphrasing of my understanding here.) So it was not that modules shouldnt have three states because that was confusing to the end-user, but because specifically *disabling but keep hanging around* was a bad idea.

So this current proposal doesnt really change this situation, to me. Or perhaps I'm missing the point?

**Comment by jenlampton** [comment]
> ...one thing to consider if we convert the checkboxes we currently have in place right now, into on/off buttons

Oh, I missed this. I don't like this idea. I vote for keeping them checkboxes, but changing the terminology to On/Off rather than enabled/disabled everywhere we refer to modules. 

> and there would be a separate page which lists "inactive" modules... Well this is exactly what we have now.

I think the recommendation was to turn the `Uninstall` page into a Report. Right now that page doesn't show any information about modules that are **not in the file system**, but still have data in the database. 

This new recommended flow makes more sense to me. Having the uninstall page be a separate action (that people need to remember) leads to people not uninstalling their module's data.

Other than that, I'm not sure a report is a huge win over the Uninstall page, except in the division of concepts, and how people will interact with different parts of the site. But I still prefer it.

I was also thinking that we could include a status report indicator (warning ?) that would alert people to the fact that they may have disabled modules that are not uninstalled, but I suppose we could add that at any time separate from this issue. So, see https://github.com/backdrop/backdrop-issues/issues/3778.

> modules could be disabled and then safely re-enabled at a later stage

I don't think this is a real use-case. Most of the time people just forget to uninstall. Either that, or they are testing something, turn off a module, and then turn it right back on again.

> So it was not that modules shouldnt have three states because that was confusing to the end-user but because specifically disabling but keep hanging around was a bad idea.

The confusing issue was part of it. Though keeping old data around is a bad idea, I think the real driver behind this change was wanting to put module state in config. Supporting that, with a disabled-but-not-uninstalled state was hard.

**Comment by docwilmot** [comment]
> @docwilmot the big change is this part:

> > On the confirmation page have two buttons - the default button turns the module off but keeps settings/data intact in case you want to turn it back on later without losing anything. The second button turns the module off and deletes all settings/data.

> This new recommended flow makes more sense to me. Having the uninstall page be a separate action (that people need to remember) leads to people not uninstalling their module's data.

I think though that this could be a regression. Currently a novice user who is "stopping" a module, would simply uncheck that box. All is well after that; a disabled module doesnt *do* anything. He doesnt need to bother with that uninstall page, really. Life would carry on just fine. Now though we would be presenting him with an additional choice to increase the anxiety - "delete all your data!!!??"

And although I suppose we could work around uninstalling multiple modules with a confirm form, thats going to be complicated. A bulk actions table with checkboxes for "just disable" vs "disable and delete all data" for *each* module being uninstalled maybe?

I understand that page is not wonderful, but I think this would complicate rather than simplify and I am not really seeing a significant gain.

**Comment by jenlampton** [comment]
> a disabled module doesn't do anything. He doesnt need to bother with that uninstall page, really. Life would carry on just fine.

This is what we have currently, only the novice user doesn't even know that uninstalling is required.

>  we would be presenting him with an additional choice to increase the anxiety - "delete all your data!!!??"

Well, we could work on that wording :) But we do need to find a way to show people that uninstalling is required. 

> And although I suppose we could work around uninstalling multiple modules with a confirm form

Or maybe a checkbox for each module? Something more like the current uninstall page, but worked into the flow so it's presented automatically.

> A bulk actions table 

I personally dislike any bulk-actions table for all the reasons previously mentioned, but I could get behind a list of checkboxes for "delete all data provided by the following X modules" 

> I think this would complicate rather than simplify 

It's not a huge increase in complication from a technical standpoint. It's essentially turning the uninstall page we have now into two different pieces, one that pops up automatically at the time modules are turned off, and one that's a report that makes it clear that something is wrong when that data is not removed. 

We can redesign, re-word, and re-work both, so that they are better suited to each task. The post-turn-off page should only list the modules that were just turned off, and all the other modules that were turned off previously (but data not deleted) will appear on the report.

The simplification is in the workflow for people using the site:
* no need to understand that a disabled module also needs to be uninstalled
* no need to remember to visit a separate page after a module has been disabled
* a helpful report page (and errors on the status report) to assist you when you forget either of the above

> and I am not really seeing a significant gain.

I personally don't see a downside. Even for the Drupal-people who are already trained to uninstall there won't be any re-training necessary because the new uninstall step would happen automatically. Any UX gain with no down-side is a win in my book :)

I've added the needs more feedback label. It would be good to get more opinions.

**Comment by klonos** [comment]
> ...we do need to find a way to show people that uninstalling is required.

I would say "recommended", or "advised"; not "required". There are use cases to keeping the configuration around (temporary disabling a module).

In order to help us understand what people use the module uninstallation routine for, and thus better solve any UX problem(s), I think that we should clarify that there are two distinct use cases:

1. **Permanently removing the module from the site** (i.e. tested and don't really like/want/need the module) -> data should be removed, but the problem is that a) we do not make that obvious for novice Drupal/Backdrop users + b) having that be a 2-step, multi-clicking-around process, increases chances that people won't be bothered doing the work. We should find a way to communicate that "if you are never going to enable this module again, and especially if you intent to remove it from your file system, then you need to uninstall (delete data) now."

2. **Temporarily disabling a module, with the intention to re-enable soon** (in order to troubleshoot or what-have-you). Point is that "temporarily" and "soon" is a few hours for example; not months/years ...and certainly not if you intent to upgrade core/modules in the meantime. If we treat this use case same as use case 1., then we are imposing to the user the need to reconfigure the module from scratch when they decide to re-enable. We should not be making such assumptions, because in this specific use case, removing module data/config without requesting confirmation (D8-style), would/could be considered data loss.

One thing that Drupal/Backdrop (any software really) does well (ironically speaking) is that we tend to "translate" in the UI, whatever happens in the code. This is plain wrong in most cases, because end-users don't necessarily know (or want/need to) what goes on in the background. In this specific use case, we are also failing to communicate in the UI that there are 2 use cases.

So I propose a confirmation page/dialog, that has 2 options:

- Red/danger button "Remove module"
- Blue/normal button "Disable module"

The help text above these buttons should be something along these lines:

> Temporarily disabling this module, keeps its configuration/data in the system, so that you have less work to do when you re-enable it later.
>
> If you are permanently uninstalling this module, then you should remove all its configuration and data, as leaving them behind can cause serious issues.
>
> To find out more, see [this b.org documentation](url).

...basically keep this message short, and to the point. Then also link to a page that explains in detail how things work "behind the scenes".

>And although I suppose we could work around uninstalling multiple modules with a confirm form

Yes, all the above will work just fine when disabling/removing individual modules. What we need to figure out is how to do this in an intuitive way when performing bulk disabling/removals. This will get tricky UI-wise in the case where you have a mix of modules that are to be disabled vs. others that are to be removed.

>Or maybe a checkbox for each module? Something more like the current uninstall page, but worked into the flow so it's presented automatically.

Yes, the easy/obvious/lazy/safe way to do the above would be to present the user with a list of the modules, and options next to each one to either remove, or disable. But imagine a list of dozens of modules ...we'd be asking the user to potentially do a lot of clicking 😓

The only alternative "clean" way I could think of right now, is to make this a two-step process:

-> user selects various modules for disabling/removal
-> we inform them that **all** of them will be _disabled_
-> user clicks "yes"
-> we present them a final dialog:
    - "modules have been successfully _disabled_" green, success message
    - "config/data should be _removed_ if not to be enabled again" dialog text
    - list of modules (no checkboxes next to them)
    - red/danger "Also remove" button
    - "skip this step" link

Basically, we'd be adding the uninstallation step in the routine; something that we currently are leaving to the end user to figure out (or know).

This leaves the use case of people needing to selectively also remove _some_ of the modules. This is where the final message comes in... when redirecting the user back to the modules listing page, we do either one of the below, depending on what the user has selected in the last step of the final "also remove things" dialog:

- Show a "all modules have been removed" green message.
- Show a "All modules have been disabled" green message + a warning: "It is advised to completely remove modules that are not used. [Link to uninstallation page](url)."

Thoughts?

**Comment by olafgrabienski** [comment]
@klonos I agree with your assumptions about use cases and also with most of the suggestions. Maybe we can integrate the use case of people needing to selectively also remove some of the modules:

-> user selects various modules for disabling/removal
-> we inform them that all of them will be disabled
-> user clicks "yes"
-> we present them a ~~final~~ dialog:
----> "modules have been successfully disabled" green, success message
----> "config/data should be removed if not to be enabled again" dialog text
----> list of modules ~~(no checkboxes next to them)~~ **with checkboxes**
----> ~~red/danger "Also remove" button~~ **user selects various modules to also remove them**
----> we inform them that all selected will be removed
----> user clicks"yes", or "Cancel"

**Comment by klonos** [comment]
@olafgrabienski I specifically wanted to avoid using checkboxes in the second step of the dialog, but perhaps if everything was already pre-ticked, this would work.

I think that it might be easier if we actually had a "live" thing to test, so I may put a draft PR together over the weekend. That way, we will be able to test and iterate on improving the UX as we get more feedback 😉.

**Comment by docwilmot** [comment]
> This is what we have currently, only the novice user doesn't even know that uninstalling is required.
"if you are never going to enable this module again, and especially if you intent to remove it from your file system, then you need to uninstall (delete data) now."
"It is advised to completely remove modules that are not used. Link to uninstallation page."

How exactly does *not* uninstalling (after disabling) impact the site?

**Comment by klonos** [comment]
>How exactly does not uninstalling (after disabling) impact the site?

I guess that there is extraneous configuration and/or data left behind @docwilmot. Basically whatever the [uninstall hooks](https://api.backdropcms.org/api/backdrop/core%21modules%21system%21system.api.php/function/hook_uninstall/1) would otherwise have taken care of I guess.

**Comment by docwilmot** [comment]
I meant in terms of performance

**Comment by jenlampton** [comment]
Backdrop builds a list of all modules and attempts to load all their files on each page load. If the module remains in the system table, but is removed from disk, Backdrop is still searching for all those files to load when they are not there. (I don't remember exactly what happens if the status in the system table is zero -- but that still caused a problem for Drupal even if the module was disabled)

This issue was resolved in Drupal 7.56 I believe, when the error message was added about modules that Drupal thought were there but were not. The problem was was much worse for Drupal than for Backdrop because Drupal also checked all possible `sites` directories automatically. In Backdrop you have to opt-in for checking those other locations.

**Comment by klonos** [comment]
> I meant in terms of performance

Oh, right 👍 ...I guess this could be listed in the documentation page we link to from the message we throw.
